### PR TITLE
Pacbio align

### LINKED ISSUE #48: Add optional read mapping subworkflow
### Description of feature
- Add read mapping flag
- Import subworkflow to map PacBio reads

### PR DESCRIPTION
**Pull request by gq1**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/sanger-tol/variantcalling/tree/main/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by gq1** [comment]
The profile `test_full_aln ` finished on the farm:

```
Completed at: 30-Nov-2023 21:33:10
Duration    : 9h 4m 3s
CPU hours   : 242.2
Succeeded   : 76
```

**Comment by gq1** [comment]
https://github.com/sanger-tol/variantcalling/issues/48

File: nextflow_schema.json [hunk 1]
```diff
@@ -28,6 +28,12 @@
                     "description": "The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure.",
                     "fa_icon": "fas fa-folder-open"
                 },
+                "vector_db": {
+                    "type": "string",
+                    "default": "/Users/gq2/git3/variantcalling/assets/vectorDB.tar.gz",
```
Comment (muffato) [inline nextflow_schema.json:33]: The default value here shouldn't be a path on your laptop :)
```suggestion
                    "default": "${projectDir}/assets/vectorDB.tar.gz",
```

**Review by muffato — COMMENTED** [review]
- Some alignment files have `.fasta` in the name, like `GCA_947369205.1_OX376310.1_CANBKR010000003.1.fasta.pacbio.icCanRufa1_combined.cram`. I guess it's because `.baseName` only removes the `.gz` extension. Can you do something like `fai.baseName - '.fasta'` (as in `subworkflows/local/input_filter_split.nf`) ?
- Not from this PR, but since we're talking about the naming convention, I would prefer not adding `_combined`
- Bump `maxRetries` to 5 in `conf/base.config`, until the resource requirements are backported from the readmapping pipeline

Thinking about the pipeline doing either the alignment sub-workflow or "samtools merge". There is actually a "samtools merge" in the alignment sub-workflow that does pretty much the same thing: align multiple aligned files from the same sample. Could there be just one "samtools merge" that works in both cases ? Also, in the alignment sub-workflow, "samtools merge" is followed by a bunch of samtools commands to extract some statistics and convert to CRAM. Here are my thoughts:

1. For resource optimisation, we'll very likely have to run "samtools flagstat" on every input file to tune CPU/memory according to the read count
2. It's maybe not a bad thing to extract all those stats from aligned files given as inputs. It may help with debugging and making sure the files are proper
3. Is CRAM conversion happening twice ? There seems to be one in `convert_stats.nf` and one in `input_filter_split.nf`. Maybe we can remove the first one ?

In summary, and I'm happy to discuss that over a call if that's easier, here is what I would propose:

1. Stop the alignment sub-workflow after minimap.
2. Run samtools merge at this stage in both alignment and pre-aligned modes.
3. The "convert stats" sub-workflow could be just "stats" and run samtools stats/flagstat/idxstat on all aligned files before input_filter_split.

**Comment by gq1** [comment]
> * Some alignment files have `.fasta` in the name, like `GCA_947369205.1_OX376310.1_CANBKR010000003.1.fasta.pacbio.icCanRufa1_combined.cram`. I guess it's because `.baseName` only removes the `.gz` extension. Can you do something like `fai.baseName - '.fasta'` (as in `subworkflows/local/input_filter_split.nf`) ?
> * Not from this PR, but since we're talking about the naming convention, I would prefer not adding `_combined`
> * Bump `maxRetries` to 5 in `conf/base.config`, until the resource requirements are backported from the readmapping pipeline
All being updated now.

**Comment by gq1** [comment]
Currently we tried to import 3 subworkflows from **readmapping** pipeline  and didn't change much.
In the future if we can push these subworkflows to nf-core, we can directly import them. 
That is why we don't need to change much for now, keep these subworkflows same way as in other pipelines (3 now?), maybe easy to maintain for now?

> Thinking about the pipeline doing either the alignment sub-workflow or "samtools merge". There is actually a "samtools merge" in the alignment sub-workflow that does pretty much the same thing: align multiple aligned files from the same sample. Could there be just one "samtools merge" that works in both cases ? Also, in the alignment sub-workflow, "samtools merge" is followed by a bunch of samtools commands to extract some statistics and convert to CRAM. Here are my thoughts:
> 
> 1. For resource optimisation, we'll very likely have to run "samtools flagstat" on every input file to tune CPU/memory according to the read count
> 2. It's maybe not a bad thing to extract all those stats from aligned files given as inputs. It may help with debugging and making sure the files are proper

I suppose the aligned input bam/cram file will come with the stats file normally.


> 3. Is CRAM conversion happening twice ? There seems to be one in `convert_stats.nf` and one in `input_filter_split.nf`. Maybe we can remove the first one ?

The first process only convert the aligned reads to CRAM and the second one will filter the reads as well.
We can skip the first one if we don't want to keep the original aligned reads.

**Comment by gq1** [comment]
The latest results here from `nextflow run variantcalling -profile test_full_align,sanger,singularity --align`

`/nfs/users/nfs_g/gq2/lustre123_gq2/git/results`

**Review by muffato — APPROVED** [review]

**Comment by muffato** [comment]
> 3\. The "convert stats" sub-workflow could be just "stats" and run samtools stats/flagstat/idxstat on all aligned files before input_filter_split.

FYI. Priyanka had the same idea for the read-mapping pipeline: https://github.com/sanger-tol/readmapping/issues/63
